### PR TITLE
Hybrid Custody contract

### PR DESCRIPTION
**Pull request by austinkline**
Second approach to the Hybrid Custody problem (still a work in progress). 

This new approach has basically three resources:
1. ChildAccount - wraps the AuthAccount capability and has helper methods to share access to the account with parents, revoke parent access, and to change ownership of the account (no owner, or delegate ownership elsewhere)
2. ProxyAccount - there is one ProxyAccount resource for every published parent. Each ProxyAccount has its own rules applied, but they can be shared across different accounts since only a capability to these rules is needed.
3. Manager - The resource used by parent accounts to control what child accounts it has redeemed, and what accounts it now owns as well. A manager can borrow a child account and then get capabilities from it

Proposed diagram (currently) looks like the following:

HybridCustody-diagram

### Current TODOs:
- [ ] Manager helper methods for owned accounts
- [ ] Helper methods to access the CapabilityProxy on a borrowed Account (public and private)

### CODE REVIEW

File: contracts/HybridCustody.cdc [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+        pub fun getCapability(path: CapabilityPath, type: Type): Capability?
+        pub fun getPublicCapability(path: PublicPath, type: Type): Capability?
+
+        access(contract) fun redeemedCallback(_ addr: Address)
+    }
+
+    // Entry point for a parent to borrow its child account and obtain capabilities or
+    // perform other actions on the child account
+    pub resource interface ManagerPrivate {
+        pub fun borrowAccount(id: UInt64): &{AccountPrivate, AccountPublic}?
```
Comment (sisyphusSmiling) [inline contracts/HybridCustody.cdc:85]: Would also be helpful to borrow an account by Address. Also wondering if we should add a method to borrow owned accounts in this interface.

**Review by sisyphusSmiling — COMMENTED** [review]
This is really great work, and a big step in the direction of developer ergonomics! Thanks for all the effort you put into this, it's clearly thoughtfully implemented.

I've added some comments and questions as points for discussion, nothing major given the high-level makes sense and will be very useful for developers building HC apps.

File: contracts/HybridCustody.cdc (hunk 1 above)
Comment (austinkline) [reply contracts/HybridCustody.cdc:85]: I think yes on both counts, here

File: contracts/HybridCustody.cdc [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+        pub fun removeChildByAddress(addr: Address)
+        pub fun removeChild(id: UInt64)
+        // TODO: Owned account methods
+    }
+
+    // Functions anyone can call on a manager to get information about an account such as
+    // What child accounts it has
+    pub resource interface ManagerPublic {
+        pub fun borrowAccountPublic(id: UInt64): &{AccountPublic}?
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:94]: Definitely, I think probably we need to have a discussion about what questions we need to be able to ask the Child or Parent accounts. I'm sure there are a few gaps still

File: contracts/HybridCustody.cdc [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+        // pub fun getChildAddresses(): [Address]
+    }
+
+    /*
+    Manager
+    A resource for an account which fills the Parent role of the Child-Parent account
+    management Model. A Manager can redeem or remove child accounts, and obtain any capabilities
+    exposed by the child account to them.
+    */
+    pub resource Manager: ManagerPrivate, ManagerPublic {
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:104]: All resources should probably implement `MetadataViews.Resolver`, not sure about `ResolverCollection` since we have default implementations now. Happy to do both, though

File: contracts/HybridCustody.cdc [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+        pub let accounts: {UInt64: Capability<&{AccountPrivate, AccountPublic}>}
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:105]: It's a good question, I didn't define a concrete type in case others want their own version of child accounts. If I define the type as `ProxyAccount{AccountPrivate, AccountPublic}`, then we will always have to stick with that type definition which might not be the right decision. We could instead add an allow list of child account types so it can never be a surprise or something?

File: contracts/HybridCustody.cdc [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+        pub let addressToAccountID: {Address: UInt64}
+
+        pub let ownedAccounts: {UInt64: Capability<&{Account, ChildAccountPrivate}>}
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:108]: Just missing, needs to be added. There is a TODO for `Owned account methods` somewhere

File: contracts/HybridCustody.cdc [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+        pub let addressToOwnedAccountID: {Address: UInt64}
+
+        pub fun addAccount(_ cap: Capability<&{AccountPrivate, AccountPublic}>) {
+            // Is there a scenario where you are shared the same address multiple times? Seems like overkill.
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:112]: We'll have to make a decision on this as a community for sure. Do we want to mandate that you cannot have overlapping addresses? Or should we allow that flexibility just in case?

File: contracts/HybridCustody.cdc [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+        }
+
+        pub fun removeChildByAddress(addr: Address) {
+            let id = self.addressToAccountID.remove(key: addr)
+                ?? panic("no child account found with the given address")
+            self.removeChild(id: id)
+        }
+
+        pub fun removeChild(id: UInt64) {
+            let cap = self.accounts.remove(key: id) ?? panic("no account found with the given id")
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:131]: See above

File: contracts/HybridCustody.cdc [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+            self.addressToOwnedAccountID.remove(key: cap.address)
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:132]: Ah yep good catch! Just a small typo

File: contracts/HybridCustody.cdc [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+
+            // TODO: emit event
+        }
+
+        pub fun addOwnedAccount(_ cap: Capability<&{Account, ChildAccountPrivate}>) {
+            let acct = cap.borrow()
+                ?? panic("cannot add invalid account")
+
+            self.ownedAccounts[acct.uuid] = cap
+            self.addressToOwnedAccountID[cap.address] = acct.uuid
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:142]: It will come down to whether we want to enforce that there are no duplicates for a parent managing two child accounts to the same address or not. If so, we would definitely want to add necessary pre-conditions like what you're suggesting

File: contracts/HybridCustody.cdc [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+
+    /*
+    The ProxyAccount resource sits between a child account and a parent and is stored on the same account as the child account.
+    Once created, a private capability to the proxy account is shared with the intended parent. The parent account
+    will accept this proxy capability into its own manager resource and use it to interact with the child account.
+
+    Because the ProxyAccount resource exists on the child account itself, whoever owns the child account will be able to manage all
+    ProxyAccount resources it shares, without worrying about whether the upstream parent can do anything to prevent it.
+    */
+    pub resource ProxyAccount: AccountPrivate, AccountPublic {
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:197]: 100%, we haven't talked about the needed MetadataViews for this yet. I think that's a whole separate discussion we should get started assuming that the general layout to this implementation is okay with everyone

File: contracts/HybridCustody.cdc [hunk 11]
```diff
@@ -0,0 +1,529 @@
...
+        // The CapabilityFilter is a restriction put at the front of obtaining any non-public Capability.
+        // Some wallets might want to give access to NonFungibleToken.Provider, but only to **some** of the collections it
+        // manages, not all of them.
+        pub let filter: Capability<&{CapabilityFilter.Filter}>
+
+        // The CapabilityProxy is a way to share one-off capabilities by the child account. These capabilities can be public OR private
+        // and are separate from the factory which returns a capability at a given path as a certain type. When using the CapabilityProxy,
+        // you do not have the ability to specify which path a capability came from. For instance, Dapper Wallet might choose to expose
+        // a Capability to their Full TopShot collection, but only to the path that the collection exists in.
+        pub let proxy: Capability<&CapabilityProxy.Proxy{CapabilityProxy.GetterPublic, CapabilityProxy.GetterPrivate}>
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:216]: Yes, exactly right 🙂

File: contracts/HybridCustody.cdc [hunk 12]
```diff
@@ -0,0 +1,529 @@
...
+            let capProxyPrivate = PrivatePath(identifier: capProxyIdentifier)!
+
+            acct.link<&CapabilityProxy.Proxy{CapabilityProxy.GetterPublic}>(capProxyPublic, target: capProxyStorage)
+            acct.link<&CapabilityProxy.Proxy{CapabilityProxy.GetterPublic, CapabilityProxy.GetterPrivate}>(capProxyPrivate, target: capProxyStorage)
+            let proxy = acct.getCapability<&CapabilityProxy.Proxy{CapabilityProxy.GetterPublic, CapabilityProxy.GetterPrivate}>(capProxyPrivate)
+            assert(proxy.check(), message: "failed to setup capability proxy for parent address")
+
+            let borrowableCap = self.borrowAccount().getCapability<&{BorrowableAccount, ChildAccountPublic}>(HybridCustody.ChildPrivatePath)
+            let proxyAcct <- create ProxyAccount(borrowableCap, factory, filter, proxy)
+            let identifier = HybridCustody.getProxyIdentifierForParent(parentAddress)
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:337]: Yeah these are a mouthful, I think we could do for an assessment of a few of the names around here, these included. I'll update these to start

File: contracts/HybridCustody.cdc [hunk 13]
```diff
@@ -0,0 +1,529 @@
...
+            let pathsToUnlink: [PrivatePath] = []
+            acct.forEachPrivate(fun (path: PrivatePath, type: Type): Bool {
+                if type.identifier == "Capability<&AuthAccount>" {
+                    pathsToUnlink.append(path)
+                }
+                return true
+            })
+
+            // Link a new AuthAccount Capability
+            let authAcctPath = "HybridCustodyRelinquished".concat(HybridCustody.account.address.toString()).concat(getCurrentBlock().height.toString())
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:460]: Yeah, that's what the note at the beginning of this method is calling out. I'll add to it and maybe copy it to this section as well to be extra safe

File: contracts/HybridCustody.cdc [hunk 14]
```diff
@@ -0,0 +1,529 @@
...
+        pub fun removeChildByAddress(addr: Address)
+        pub fun removeChild(id: UInt64)
+        // TODO: Owned account methods
+    }
+
+    // Functions anyone can call on a manager to get information about an account such as
+    // What child accounts it has
+    pub resource interface ManagerPublic {
+        pub fun borrowAccountPublic(id: UInt64): &{AccountPublic}?
+        // pub fun getChildAddresses(): [Address]
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:97]: Just not implemented. I often times will define and comment out the methods I want on an interface to try and not forget about them 🙂

File: contracts/HybridCustody.cdc (hunk 12 above)
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:366]: Changed to `getProxyAccountIdentifier` and renamed the other helper methods as well

File: contracts/HybridCustody.cdc (hunk 3 above)
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:106]: going to add a TODO here so we can address it in a later PR

File: contracts/HybridCustody.cdc (hunk 13 above)
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:494]: Added an extra note here, we can address when we get to capability controllers

File: contracts/HybridCustody.cdc [hunk 15]
```diff
@@ -0,0 +1,529 @@
...
+            self.addressToAccountID[cap.address] =  acct.uuid
+            
+            // TODO: emit account registered event
+
+            acct.redeemedCallback(self.owner!.address)
+        }
+
+        pub fun removeChildByAddress(addr: Address) {
+            let id = self.addressToAccountID.remove(key: addr)
+                ?? panic("no child account found with the given address")
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc]: Went with a compromise, here. Added a TODO to emit an event only if an item was actually removed that way we can ensure managers can always remove things. We don't want to let child accounts mess with or invalidate the manager address so hopefully events is enough of a signal.

File: contracts/HybridCustody.cdc (hunk 6 above)
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/HybridCustody.cdc:115]: Added pre checks here

File: contracts/HybridCustody.cdc (hunk 10 above)
[the remark this answers is not in the document]
Comment (sisyphusSmiling) [reply contracts/HybridCustody.cdc:198]: That makes sense! You're right, the views can be layered on after this is merged. I just made #18 for that topic, so it's captured somewhere.

File: contracts/HybridCustody.cdc [hunk 16]
```diff
@@ -0,0 +1,533 @@
...
+    // A ChildAccount shares the BorrowableAccount capability to itelf with ProxyAccount resources
+    pub resource interface BorrowableAccount {
+        access(contract) fun borrowAccount(): &AuthAccount
+    }
+
+    // Public methods anyone can call on a child account
+    pub resource interface ChildAccountPublic {
+        pub fun getParentsAddresses(): [Address]
+        pub fun getParentStatuses(): {Address: Bool}
+        pub fun getRedeemedStatus(addr: Address): Bool?
```
Comment (joshuahannan) [inline contracts/HybridCustody.cdc:50]: Can you add some comments explaining what a parent status and redeemed status is?

File: contracts/HybridCustody.cdc [hunk 17]
```diff
@@ -0,0 +1,533 @@
...
+    // Accessible to the owner of the ChildAccount.
+    pub resource interface ChildAccountPrivate {
+        pub fun removeParent(parent: Address): Bool
+        pub fun publishToParent(
+            parentAddress: Address,
+            factory: Capability<&CapabilityFactory.Manager{CapabilityFactory.Getter}>,
+            filter: Capability<&{CapabilityFilter.Filter}>
+        )
+        pub fun giveOwnership(to: Address)
+        pub fun seal()
```
Comment (joshuahannan) [inline contracts/HybridCustody.cdc:64]: Can you add a comment about what `seal()` means?

File: contracts/HybridCustody.cdc [hunk 18]
```diff
@@ -0,0 +1,533 @@
...
+        pub let ownedAccounts: {Address: Capability<&{Account, ChildAccountPrivate}>}
+
+        pub fun addAccount(_ cap: Capability<&{AccountPrivate, AccountPublic}>) {
+            pre {
+                self.accounts[cap.address] == nil: "There is already a child account with this address"
+            }
+
+            // Is there a scenario where you are shared the same address multiple times? Seems like overkill.
+            let acct = cap.borrow()
+                ?? panic("invalid account capability")
```
Comment (joshuahannan) [inline contracts/HybridCustody.cdc:117]: maybe add a more descriptive error message here

File: contracts/HybridCustody.cdc [hunk 19]
```diff
@@ -0,0 +1,533 @@
...
+    ChildAccount
+    A resource which sits on the account it manages to make it easier for apps to configure the behavior they want to permit.
+    A ChildAccount can be used to create ProxyAccount resources and share them publish them to other addresses.
+
+    The ChildAccount can also be used to pass ownership of an account off to another address, or to relinquish ownership entirely,
+    marking the account as owned by no one. Note that even if there isn't an owner, the parent accounts would still exist, allowing
+    a form of Hybrid Custody which has no true owner over an account, but shared partial ownership.
+    */
+    pub resource ChildAccount: Account, BorrowableAccount, ChildAccountPublic, ChildAccountPrivate {
+        pub var acct: Capability<&AuthAccount>
```
Comment (joshuahannan) [inline contracts/HybridCustody.cdc:283]: maybe it would be good to make this private?

**Review by joshuahannan — COMMENTED** [review]
Looks pretty good to me! Mostly small comments about docs, error messages, and such. I agree that we should add more events and metadata views stuff

File: contracts/CapabilityFilter.cdc [hunk 20]
```diff
@@ -31,7 +31,7 @@ pub contract CapabilityFilter {
         }
 
         pub fun allowed(cap: Capability): Bool {
-            return self.deniedTypes[cap.borrow<&AnyResource>().getType()] == nil
+            return !self.deniedTypes.containsKey((cap.borrow<&AnyResource>() ?? panic("requested type is not allowed")).getType())
```
[the remark this answers is not in the document]
Comment (austinkline) [reply contracts/CapabilityFilter.cdc]: Oversight! Too hung up on trying to not use gas by assigning a new variable. Cleaned this up with an `if let ... `

File: contracts/HybridCustody.cdc (hunk 17 above)
Comment (austinkline) [reply contracts/HybridCustody.cdc:64]: Added a comment. We might want to remove this from the interface, though. It might only be relevant for this implementation of ChildAccount, thoughts?

**Comment by austinkline** [comment]
> Looks pretty good to me! Mostly small comments about docs, error messages, and such. I agree that we should add more events and metadata views stuff

Thanks for taking a look! My hope is that we can form breakout discussions to quickly iterate on metadata views and on events as soon as this is merged. If we are okay with the general approach to access, I think we can let this go pretty soon and then start flagging what's left for anyone else to come in and contribute if they want

File: contracts/HybridCustody.cdc (hunk 1 above)
Comment (austinkline) [reply contracts/HybridCustody.cdc]: Added these methods, but added a comment about what we should be able to obtain from an owned account. We might need a CapabilityFactory instance there as well for the owner to maintain what you can borrow straight from there (there is no proxy in that scenario.)

**Comment by austinkline** [comment]
Last main piece I think we should tackle here:
1. How to let the owner manage a child account's CapabilityFilter, and CapabilityProxy
2. What we need to be able to discover on a child account from the owned account

File: contracts/HybridCustody.cdc (hunk 17 above)
Comment (sisyphusSmiling) [reply contracts/HybridCustody.cdc:81]: I agree, feels like adding it to the interface would be overly prescriptive. It's also only used in `giveOwnership()` in this implementation, so removing it from the interface wouldn't affect intended use here

Comment (austinkline) [reply contracts/HybridCustody.cdc:81]: It's used by `giveOwnership` but also can be used on its own. This basically expresses two flavors of ownership:
1. The current owner transfers ownership of the account away to a new address
2. The current owner decides no one should own the account, and seals the account so that no one can add or remove any future resources, all they can do is interact with what's already there.

The second flavor of ownership is a much more trust-less approach where no one can alter the underlying paths or capabilities that are relied on. We want a way to let the owner exercise that function, but probably can find another way to expose it to the owner

**Comment by sisyphusSmiling** [comment]
> How to let the owner manage a child account's CapabilityFilter, and CapabilityProxy

Once they're the owner, they would have access to all resources in the child account, giving them the ability to add arbitrary Capabilities to any CapabilityProxy in said account.

Thinking of the case @rrrkren mentioned where wallets like Dapper Wallet have restricted access on a child account on which another self-custodial wallet has owner access, this raises an issue. For instance, DW would not want to have an access path on a FungibleToken Capability without their authorization on that access path.

This leads me to believe that access should either be immutable or require a handshake of sorts to grant new Capabilities. If I'm an owner and I want to give a parent access to an additional capability, I should ask them to receive that Capability.

Put simply, owners should be able to initiate adding capabilities but require parent approval to do so while reserving sole authority to remove capabilities from a proxy or filter.

One solution here might be removing the ability to add Capabilities to proxies and filters altogether. This implies the need to create a new `ProxyAccount` and replace the corresponding Capability in the parent's `Manager` for the associated `ChildAccount`.

Curious to hear others' thoughts here.

> What we need to be able to discover on a child account from the owner

Thoughts here are:
- All associated parent accounts
- Metadata about the account

I would say the contents of the child account, but that can be done via scripts/off-chain systems as is already common practice.

File: contracts/HybridCustody.cdc (hunk 17 above)
Comment (sisyphusSmiling) [reply contracts/HybridCustody.cdc:95]: I see, that makes sense. Great idea. I don't think it's a problem to keep it in the interface

**Comment by austinkline** [comment]
> Once they're the owner, they would have access to all resources in the child account, giving them the ability to add arbitrary Capabilities to any CapabilityProxy in said account.
> 
> Thinking of the case @rrrkren mentioned where wallets like Dapper Wallet have restricted access on a child account on which another self-custodial wallet has owner access, this raises an issue. For instance, DW would not want to have an access path on a FungibleToken Capability without authorizing their being granted that access.
> 
> This leads me to believe that access should either be immutable or require a handshake of sorts to grant new Capabilities. If I'm an owner and I want to give a parent access to an additional capability, I should ask them to receive that Capability.
> 
> Put simply, owners should be able to initiate adding capabilities but require parent approval to do so while reserving sole authority to remove capabilities from a proxy or filter.
> 
> One solution here might be removing the ability to add Capabilities to proxies and filters altogether. This implies the need to create a new `ProxyAccount` with new `CapabilityFilter` & `CapabilityProxy` and replace the corresponding Capability in the parent's `Manager` for the associated `ChildAccount`.

We want this management back and forth to be as light a lift as possible. If we need to put either of these options in place, we should have a discussion about it to be sure that this is necessary since they will make the control flow here much more complex. It could be that Dapper needs a custom manager resource which lets them gate what capabilities they can get regardless of what the underlying child account does

**Comment by austinkline** [comment]
Had some [discussion offline in the flow discord](https://discord.com/channels/613813861610684416/1087374662100602920/1101670907568730152), we're going to look into a mechanism that lets the manager gate what types can be returned by the proxy and factory, regardless of what's in the ProxyAccount resource. We'll want confirmation from the Dapper Wallet team that this approach would suffice, but it sounds like a workable approach to ensure parent accounts can follow whatever rules they have to adhere to.

The plan is to tackle that and discoverability questions in separate PRs so we can keep this moving forward. There is certainly much work to do, still, but if we stall too long nothing will happen and we'll be in the same position we are now in a month. 

I am planning to merge this PR by EOD today so we can proceed with next steps. Mainly:
1. Events for indexer discoverability
2. Metadata Views
3. Discoverability helper functions. We need to explore what questions you should be able to ask the Child, Parent, or in some cases the Owner. 

Each of these I think we can tackle separately but it's possible the MetadataViews piece will be closely tied to the discoverability piece. I'm guessing we will leverage metadata views for discoverability.

**Review by sisyphusSmiling — APPROVED** [review]
LGTM! On board with the plan to parallelize work on the topics you mentioned. If we could add getters for CapabilityProxy in this is PR as [mentioned here](https://github.com/Flowtyio/restricted-child-account/pull/14/files#r1175815839), that would be great. But not opposed to adding that in future PRs to get this merged ASAP. Thanks for working hard to keep things rolling!

**Review by sisyphusSmiling — DISMISSED** [review]
LGTM!

**Comment by austinkline** [comment]
Retrieval through the Capability Proxy is now working as a first version, and the manager resource restricting what's returning in private types works now. I wasn't sure if we needed to also restrict public types, the thought behind not applying a manager's CapabilityFilter to public capabilities via the proxy or factory is that it's still accessible to them because it would be accessible to _everyone_. Hopefully that's the right approach, we will want to confirm.

**Comment by austinkline** [comment]
I think that's it for this version! @sisyphusSmiling feel free to review again if you would like, I will merge this later tonight and we can plan next steps tomorrow 🤝
